text/0069-schema-libraries.md

```diff
@@ -0,0 +1,178 @@
+  either library definitions for use in schemas or library functions for use in
+  Cedar authorization calls. It only points out that the community could
+  converge on these things.)
+
+## Detailed design
+
+Import statements are only allowed at the top level, outside of all namespace
+declarations.
+(In the future, another RFC could propose allowing imports in other positions.)
+
+The target of the import must be a raw file containing a valid Cedar schema.
+This schema may contain any definitions that are valid today in Cedar schemas,
+including namespaces, entity type definitions, common type definitions, and
+action declarations.
+The validator will essentially concatenate all of these definitions into the
+schema at the location of the `import` statement.
+
+Cedar will autodetect whether the imported schema is a human-format or
+JSON-format schema.
+(Today, there are no strings that are both valid human-format and valid
+JSON-format schemas; this RFC proposes encoding that as a design principle in
+perpetuity.)
+In particular, all valid JSON-format schemas must have `{` as their first
+non-whitespace character, and no valid human-format schemas have `{` as their
+first non-whitespace character.
+
+This RFC does not propose any mechanism for versioning libraries.
+Instead, it proposes that versioning would be done _above_ the Cedar layer,
+i.e., should be the responsibility of library authors.
+For instance, library authors could provide a different URL for different
+versions of their library, avoiding changing the contents of the URL for the
+existing versions of the library.
+This RFC doesn't preclude later adding a versioning feature, in which case the
+syntax proposed in this RFC would be interpreted as "import the latest version
+of this library".
+
+In the JSON format, we do not need to reserve the namespace named `"imports"`:
+if `"imports"` maps to a JSON object, it represents the namespace `"imports"`,
+while if `"imports"` maps to a JSON array, it represents import statements as
+defined in this RFC.
+
+## Drawbacks
+
+1. The Cedar validator, and other tools that rely on schemas, will have to make
+network calls in order to perform their jobs. This has availability and latency
+implications which may not be acceptable for some users. Of course, those users
+could simply not use this feature.
+2. Cedar schemas would no longer be self-contained, in that a single (hopefully
+readable) file contains all of the relevant definitions. To mitigate this, we
+could provide a utility that displays the schema with all imports expanded.
+3. The Cedar Rust code would have to bring in substantial new dependencies, so
+that it could download libraries from remote URLs. To mitigate this for users
+who are concerned about this and don't need/want this feature (e.g., in
+resource-constrained environments, offline environments, Wasm, etc), we could
+put this RFC's functionality behind a Cargo feature, so that it and its
```

Comment:
Possibly local file system dependency enabled by default, with network dependencies behind a flag?

Reply:
Your comment makes me realize I should definitely explicitly call out allowing local file system URIs in addition to URLs, which is not currently made explicit in the RFC. I'll amend

Reply:
Take a look at what IonSchema does with their "SchemaAuthority": https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec. Essentially decoupling the loading of dependencies from their declaration. 

But I'm wondering if we really need/want to bring this into Cedar. (See my other comment.)